[결제] 사전충전 기능 구현

zmfpdl64 opened the pull request:
## 개요
```mermaid
sequenceDiagram
Client ->> Payment API: 사전 충전 금액 기록
Payment API ->> Client: 충전 ID반환
```
Client에서 토스 위젯을 불러오기 전에 서버로 사전 충전 금액을 전달합니다.
this branch: #33 

## PR 유형
어떤 변경 사항이 있나요?

- [x] 새로운 기능 추가
- [x] 주석 추가 및 수정
- [x] 문서 수정
- [x] 테스트 추가, 테스트 리팩토링
- [x] 빌드 부분 혹은 패키지 매니저 수정

## PR Checklist
PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.  Commit message convention 참고  (Ctrl + 클릭하세요.) 
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

> ### Description

- [x] 시퀸스 다이어그램 작성
- [x] 컨트롤러 작성
- [x] 서비스 작성
- [x] DTO 작성
- [ ] 예외 작성(다른 API 연동시 개발)
- [x] 테스트 코드 작성


> ### Core Code

```java
	@Transactional
	public BeanPayDto.Response preChargeBeanPay(final BeanPayDto.Request.PreCharge request) {
		// TODO: 유저 금액있는지 검증로직 구현예정
		
		final BeanPay beanPay = BeanPay.ofCreate(request.userId(), request.amount());
		final BeanPay createBeanPay = beanPayRepository.save(beanPay);

		return BeanPayDto.Response.ofCreate(createBeanPay);
	}
```

zmfpdl64 committed:
Chore: #33 빈페이 사전결제 설정파일 작성

zmfpdl64 committed:
Chore: #33 빈페이 사전결제 설정파일 작성

zmfpdl64 committed:
Refactor: #33 빈페이 사전결제 클래스 수정

zmfpdl64 committed:
Refactor: #33 빈페이 사전결제 클래스 수정

zmfpdl64 committed:
Feat: #33 사전충전객체를 생성하기 위한 기능 개발

zmfpdl64 committed:
Feat: #33 사전충전객체를 생성하기 위한 기능 개발

zmfpdl64 committed:
Test: #33 사전충전객체를 생성하기 위한 기능 테스트
컨트롤러
서비스
레포지토리
DTO
Entity

zmfpdl64 committed:
Test: #33 사전충전객체를 생성하기 위한 기능 테스트
컨트롤러
서비스
레포지토리
DTO
Entity

Joowon-Seo commented on payment-api/src/main/java/org/ecommerce/paymentapi/controller/PaymentController.java:
```diff
@@ -15,10 +20,13 @@
 	private final PaymentServiceImpl paymentService;
+	private final BeanPayService beanPayService;
 
-	@GetMapping
-	public Response<Void> testApi() {
-		paymentService.createPayment();
-		return new Response<>(200, null);
+	@GetMapping("/beanpay/pre-charge")
+	public Response<BeanPayDto.Response> preCharge(@RequestBody final BeanPayDto.Request.PreCharge request) {
+		final BeanPayDto.Response preCharge = beanPayService.preChargeBeanPay(request);
+		return new Response<>(200, preCharge);
 	}
+
+
 }
```
지극히 주관적인 생각입니다!
도메인은 beanpay이고, 안에서의 행위가 충전이라 `beanpay/payments`의 구조가 조금 더 자연스러워 보입니다.

zmfpdl64 replied:
안녕하세요 주원님 의견 감사합니다. payment MS 서버 안에 결제와 빈페이 도메인이 있다고 생각해서 저런 구조를 작성했던것 같습니다. 주원님 의견에 공감하고 반영해보도록 하겠습니다 

pre-charge 경로는 행위를 보여주기 때문에 제거하는게 맞다고 생각이 드네요 조언 감사합니다 :)

Joowon-Seo replied:
```
@RequestMapping("/beanpay")
beanpayController{
@Getmapping("/payment")
}
```
공감해주셔서 감사합니당!!
혹시 제 의견이 잘못 전달됐을까봐 예시 코드 남깁니다!

zmfpdl64 replied:
예시코드까지 작성해주셨군요!! 맞습니다 이렇게 Beanpay 컨트롤러와 PaymentController로 분리해서 역할과 책임을 분리했습니다! 감사합니다

https://github.com/Team-Koreano/Koreano/pull/42/commits/61f76ddf8bd7d61a9086ca6e8fc1b7cb5fb7e664

Joowon-Seo reviewed (approved):
`implementation 'org.springframework.cloud:spring-cloud-starter-openfeign` 벌써 준비 중 이시군요.

고생하셨습니다!

hjm12223 reviewed (approved):
수고하셨습니다!

zmfpdl64 committed:
Refactor: 컨트롤러 도메인 분리

Payment MS안에 Payment 컨트롤러 안에 결제 기능과 빈페이 기능을 구현하는 것보다는 각각의 도메인별로 컨트롤러를 분리하는게 유지보수성에도 좋아보임

zmfpdl64 committed:
Refactor: 컨트롤러 도메인 분리

Payment MS안에 Payment 컨트롤러 안에 결제 기능과 빈페이 기능을 구현하는 것보다는 각각의 도메인별로 컨트롤러를 분리하는게 유지보수성에도 좋아보임

no-oneho reviewed (approved):
테스트까지 구체적으로 잘 작성하신것 같습니다! 고생하셨습니다.

zmfpdl64 committed:
Merge remote-tracking branch 'origin/develop' into feature/payment/#33

# Conflicts:
#	build.gradle

zmfpdl64 committed:
Refactor: build.gradle conflict 해결

zmfpdl64 committed:
Fix: build.gradle 코드 수정

zmfpdl64 committed:
Merge branch 'feature/payment/#33' of https://github.com/Team-Koreano/Koreano into feature/payment/#33

# Conflicts:
#	build.gradle

zmfpdl64 committed:
Fix: valid 의존성 추가
